### PR TITLE
example quantum anomaly detection

### PR DESCRIPTION
**Pull request by Simone-Bordoni**
To be completed, 
I will add README, tests and .ipynb

Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Comment by Simone-Bordoni** [comment]
I need some help with tests that are not passing.
When the script is run locally everythong works.

**Comment by scarrazza** [comment]
From the last crash, seems like you are writing to a folder that does not exist: `results/loss_distribution.png`.

**Review by MatteoRobbiati — CHANGES_REQUESTED** [review]
Thank you for this, it is in a good shape for me. 
I left in the review some small comments, many of which are really small fixes. I suggest two larger changes:

1. Reverse the location of the "How to run an example" section with the theoretical section in `README.md`. I think it's more natural to read the theoretical background and then have in sequence: explanation of how to execute, anticipation of the results to be derived.
2. Add docstrings in the codes so that the implemented functions are clear when reading.

File: examples/anomaly_detection/README.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# Anomaly detection with variational quantum circuits
+
+Code at: [https://github.com/qiboteam/qibo/tree/master/examples/anomaly_detection](https://github.com/qiboteam/qibo/tree/master/examples/anomaly_detection).
+
+## Problem overview
+
+With the contemporary peak in interest regarding machine learning algorithms for their many applications in scientific research as well as industrial technology, we also have the simultaneous development of quantum computing. A combination of these two fields has lead to the development of quantum machine learning algorithms.
+With this example we want to study the quantum version of a classic machine learning algorithm known as anomaly detection. This algorithm is implemented with an artificial neural network, in particular an autoencoder. In quantum machine learning, the autoencoder is realised using a variational quantum circuit. The proposed algorithm is not meant to outperform the classical counterpart on classical data. This work aims to demonstrate that it is possible to use quantum variational algorithms for anomaly detection with possible future advantages in the analysis of quantum data.
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/README.md:8]:
```suggestion
With this example we want to study the quantum version of a classic machine learning algorithm known as [anomaly detection](https://arxiv.org/abs/2007.02500). This algorithm is implemented with an artificial neural network, in particular an autoencoder. In quantum machine learning, the autoencoder is realised using a variational quantum circuit. The proposed algorithm is not meant to outperform the classical counterpart on classical data. This work aims to demonstrate that it is possible to use quantum variational algorithms for anomaly detection with possible future advantages in the analysis of quantum data.
```

File: examples/anomaly_detection/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+## Background
+
+In this section we want to explain the main elements to understand the proposed quantum anomaly detection algorithm.
+
+### Anomaly detection
+
+Anomaly detection is a classification algorithm that allows to identify anomalous data. The advantage in using this machine learning technique is that only a dataset with non anomalous (standard) data samples is required for the training.
+To achieve this it's necessary to train a particular artificial neural network (ANN) architecture called autoencoder. An autoencoder is composed of two main parts: encoder and decoder.
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/README.md:17]:
```suggestion
To achieve this it's necessary to train a particular artificial neural network (ANN) architecture called [autoencoder](https://arxiv.org/abs/2003.05991). An autoencoder is composed of two main parts: encoder and decoder.
```

File: examples/anomaly_detection/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+It is possible to define the following hyper-parameters for the performance evaluation of the circuit, `n_layers` must be equal to the one used for training:
+- `n_layers` (int): number of ansatz circuit layers (default 6).
+- `train_size` (int): number of samples used for training, the remainings are used for performance evaluation, total samples are 7000 (default 5000).
+- `filename` (str): location and file name of trained parameters to be tested (default "parameters/trained_params.npy").
+- `plot` (bool): make plots of ROC and loss function distribution (default True).
+- `save_loss` (bool): save losses for standard and anomalous data (default False).
+
+As an example, in order to use 4 layers in the variational quantum ansatz, you should execute the following command for training:
+
+```python
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/README.md:82]:
```suggestion
```bash
```

File: examples/anomaly_detection/README.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+python train.py --n_layers 4
+```
+And the following command for performance evaluation:
+
+```python
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/README.md:87]:
```suggestion
```bash
```

File: examples/anomaly_detection/test.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        """Create encoder quantum circuit.
+
+        Args:
+            n_qubits (int): number of qubits in the circuit.
+            n_layers (int): number of ansatz circuit layers.
+            params (tf.Tensor): parameters of the circuit.
+            q_compression (int): number of compressed qubits.
+
+        Returns:
+            encoder (qibo.models.Circuit): variational quantum circuit.
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/test.py:36]:
```suggestion
            encoder (:class:`qibo.models.circuit.Circuit`): variational quantum circuit.
```

File: examples/anomaly_detection/test.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            encoder.add(gates.RZ(j, params[index + 2]))
+            index += 3
+        return encoder
+
+    # Evaluate loss function (3 qubit compression) for one sample
+    def compute_loss_test(encoder, vector):
+        """Evaluate loss function for one test sample.
+
+        Args:
+            encoder (qibo.models.Circuit): variational quantum circuit (trained).
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/test.py:63]:
```suggestion
            encoder (:class:`qibo.models.circuit.Circuit`): variational quantum circuit (trained).
```

File: examples/anomaly_detection/train.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        """Create encoder quantum circuit.
+
+        Args:
+            n_qubits (int): number of qubits in the circuit.
+            n_layers (int): number of ansatz circuit layers.
+            params (tf.Variable): parameters of the circuit.
+            q_compression (int): number of compressed qubits.
+
+        Returns:
+            encoder (qibo.models.Circuit): variational quantum circuit.
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/train.py:38]:
```suggestion
            encoder (:class:`qibo.models.circuit.Circuit`): variational quantum circuit.
```

File: examples/anomaly_detection/train.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            index += 3
+        return encoder
+
+    # Evaluate loss function (3 qubit compression) for one sample
+    @tf.function
+    def compute_loss(encoder, params, vector):
+        """Evaluate loss function for one train sample.
+
+        Args:
+            encoder (qibo.models.Circuit): variational quantum circuit.
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/train.py:66]:
```suggestion
            encoder (:class:`qibo.models.circuit.Circuit`): variational quantum circuit.
```

File: examples/anomaly_detection/train.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+        return loss
+
+    # One optimization step
+    @tf.function
+    def train_step(batch_size, encoder, params, dataset):
+        """Evaluate loss function on one train batch.
+
+        Args:
+            batch_size (int): number of samples in one training batch.
+            encoder (qibo.models.Circuit): variational quantum circuit.
```
Comment (MatteoRobbiati) [inline examples/anomaly_detection/train.py:91]:
```suggestion
            encoder (:class:`qibo.models.circuit.Circuit`): variational quantum circuit.
```

**Review by MatteoRobbiati — CHANGES_REQUESTED** [review]
Thank you for your work.

Some more suggestions follow: some more references.
If you like them, you can apply the modifications directly by adding the suggestions I purpose.

**Comment by MatteoRobbiati** [comment]
I have done a last small fix, now I think it can be merged.

**Review by scarrazza — APPROVED** [review]
Looks good, I have linked the new example in sphinx.
